paddle/cinn/ir/schedule_block_graph.h

```diff
@@ -0,0 +1,200 @@
+  // is being considered.
+  std::vector<Expr> ControlStmts() const;
+
+  // Get all the upstream nodes that this node depends on.
+  std::unordered_set<std::string> UpstreamNodes() const {
+    return upstream_nodes_;
+  }
+
+  // Get all downstream nodes that depend on this node.
+  std::unordered_set<std::string> DownstreamNodes() const {
+    return downstream_nodes_;
+  }
+
+  // Get the producer node that this node directly depends on
+  std::vector<ScheduleBlockNode*> Producers() const;
+
+  // Get consumer nodes that directly depend on this node.
+  std::vector<ScheduleBlockNode*> Consumers() const;
+
+  void AddUpstreamNode(const std::string& node_id) {
+    upstream_nodes_.insert(node_id);
+  }
+  void AddDownstreamNode(const std::string& node_id) {
+    downstream_nodes_.insert(node_id);
+  }
+
+ private:
+  std::vector<common::Shared<common::GraphEdge>> OrderedInLinks() const;
+  std::vector<common::Shared<common::GraphEdge>> OrderedOutLinks() const;
+
+ private:
+  std::string id_;
+  std::unordered_set<std::string> upstream_nodes_;
+  std::unordered_set<std::string> downstream_nodes_;
+  const IRSchedule& ir_sch_;
+};
+
+// Graph in units of ScheduleBlockNode, each node corresponds to a ScheduleBlock
+// in IR.
+class ScheduleBlockGraph : public common::Graph {
+ public:
+  explicit ScheduleBlockGraph(const IRSchedule& ir_sch);
+
+  // Update graph information according to the new IRSchedule.
+  void Update(const IRSchedule& ir_sch);
+
+  // Retrieve a node in the graph by id, the id is same as the name of
+  // ScheduleBlock.
+  ScheduleBlockNode* RetrieveNode(const std::string& id) {
+    return dynamic_cast<ScheduleBlockNode*>(common::Graph::RetrieveNode(id));
+  }
+
+  // Get all block name in order,
+  // this sequence may become invalid after some schedule operations,
+  // and an Update() operation is required.
+  std::list<std::string> BlockIdsInOrder() const { return block_ids_in_order_; }
+
+  // Get all nodes without input node.
+  std::vector<ScheduleBlockNode*> StartPoints();
+
+  // Get all nodes without output node.
+  std::vector<ScheduleBlockNode*> EndPoints();
+
+  // Function used to define the operations to be performed on each node.
+  using NodeHandlerType = std::function<void(ScheduleBlockNode*)>;
+
+  // Walk through each node
+  // and perform some operations defined by NodeHandler on it.
+  void NodesWalk(const NodeHandlerType& NodeHandler);
+
+  // Walk through each node topological dfs topo order
+  // and perform some operations defined by NodeHandler on it.
+  void DFSTopoWalk(const NodeHandlerType& NodeHandler, bool is_reverse = true);
+
+ private:
+  std::list<std::string> block_ids_in_order_;
+};
+
+/**
+ * The mutator used to construct the order of blocks and their control
+ * statements
+ *
+ * Example:
+ * for0:
+ *   for1:
+ *     block0
+ *     block1
+ *   block2
+ *   for2:
+ *     block3
+ *     block4
+ *
+ * the result is:
+ *   [0]: for0
```

Comment:
Just reminder, not comment, I guess you are making some marks for control structures.

1. Is integer enough for your purpose? Are you going to distinguish for\block
2. Should brackets block (our ir::Block, not ScheduleBlock) be consider? In C++, it will be like:
```
int a = ...
{  // brackets block begins
   int a = ... // a has different meaning and life cycle but only in brackets block
}  // brackets block ends
```

Reply:
I attempted to obtain the order of the ScheduleBlock, For, and IfThenElse nodes in the entire IR, which have a mutual inclusion relationship.
I constructed a map with std::vector as the key, similar to the titles used in writing articles, with different levels, and each title under same level can be sorted.

1.Therefore, the three types of nodes in this trial have the same attributes and do not need to be distinguished.
2.In addition, the content under the three types of nodes mentioned above can be considered as Blocks, with only one more parenthesis at this point. Currently, we may not consider it.

Reply:
Just suggestion/reminder :-)

Reply:
Currently, there is no need to differentiate between 3 types. If there is a need in the future, further improvements will be made.